testing/opae-c/test_init_c.cpp

```diff
@@ -166,3 +172,245 @@ TEST(init, log_file) {
+    cfg_stream.close();
+
+    setenv("WITH_ASE", "1", 0);
+  }
+
+  virtual void TearDown() override {
+    unsetenv("WITH_ASE");
+    unlink(cfg_file_.c_str());
+    // remove any directories we created in SetUp
+    while (!dirs_.empty()) {
+      unlink(dirs_.top().c_str());
+      dirs_.pop();
+    }
+    // restore the opae_ase.cfg file at OPAE_ASE_CFG_SRC_PATH
+    rename(src_cfg_file_.c_str(), (char *)OPAE_ASE_CFG_SRC_PATH);
+  }
+
+  char buffer_[PATH_MAX];
+  std::string cfg_file_;
+  char *cfg_dir_;
+  std::stack<std::string> dirs_;
+  std::string src_cfg_file_;
+};
+
+
+/**
+ * @test       find_ase_cfg_2
+ * @brief      Test: find_ase_cfg at OPAE_ASE_CFG_INST_PATH and OPAE_PLATFORM_ROOT
+ *             release location <br>
+ * @details    After renaming a configuration file located in the OPAE_ASE_CFG_SRC_PATH
+ *             and OPAE_ASE_CFG_INST_PATH<br>
+ *             When I call find_ase_cfg
+ *             Then the call is successful<br>
+ */
+TEST_P(init_ase_cfg_p, find_ase_cfg_2) {
+    char *cfg_file = nullptr;
+    std::string inst_cfg_file_;
+
+    // find_ase_cfg at OPAE_ASE_CFG_INST_PATH
+    cfg_file = find_ase_cfg();
+    EXPECT_NE(cfg_file, nullptr);
+
+    // copy it to a temporary buffer that we can use dirname with
+    std::string inst_cfg_path = (OPAE_ASE_CFG_INST_PATH? OPAE_ASE_CFG_INST_PATH : "");
+    std::copy(inst_cfg_path.begin(), inst_cfg_path.end(), &buffer_[0]);
+    char *inst_cfg_dir = dirname(buffer_);
+    char *inst_cfg_file = basename(buffer_);
+
+   // rename opae_ase.cfg under releae directory
+    std::string cfg_dir = (inst_cfg_dir? inst_cfg_dir : "");
+    inst_cfg_file_ = cfg_dir + std::string(inst_cfg_file) + std::string(".backup");
+    rename(OPAE_ASE_CFG_INST_PATH, inst_cfg_file_.c_str());
+
+    // find_ase_cfg at release directory
+    cfg_file = find_ase_cfg();
+    EXPECT_NE(cfg_file, nullptr);
+    rename(inst_cfg_file_.c_str(), OPAE_ASE_CFG_INST_PATH);
+}
+
+/**
+ * @test       find_ase_cfg_3
+ * @brief      Test: find_ase_cfg at HOME location
+ * @details    After renaming a configuration file located in the OPAE_ASE_CFG_SRC_PATH,
+ *             OPAE_ASE_CFG_INST_PATH and OPAE release directory<br>
+ *             When I call find_ase_cfg
+ *             Then the call is successful<br>
+ */
+TEST_P(init_ase_cfg_p, find_ase_cfg_3) {
+    char *cfg_file = nullptr;
+    char *opae_path;
+    std::string  inst_cfg_file_, opae_cfg_file_;
+    std::string  rel_cfg_file_, rel_cfg_file2_;
+
+    // copy it to a temporary buffer that we can use dirname with
+    std::string inst_cfg_path = (OPAE_ASE_CFG_INST_PATH? OPAE_ASE_CFG_INST_PATH : "");
+    std::copy(inst_cfg_path.begin(), inst_cfg_path.end(), &buffer_[0]);
+    char *inst_cfg_dir = dirname(buffer_);
+    char *inst_cfg_file = basename(buffer_);
+
+    std::string cfg_dir = (inst_cfg_dir? inst_cfg_dir : "");
+    inst_cfg_file_ = cfg_dir + std::string(inst_cfg_file) + std::string(".backup");
```

Comment:
What happens if cfg.backup already existed?

Reply:
Will be overwritten.

Reply:
Use mkstemp instead.

Reply:
done